application/comit_node/src/comit_i_routes.rs

```diff
@@ -0,0 +1,98 @@
+use comit_i::Asset;
+use http::{uri::PathAndQuery, Response, StatusCode};
+use mime_guess::Mime;
+use std::{
+    borrow::Cow,
+    error::Error as StdError,
+    fmt::{self, Display},
+    str::FromStr,
+};
+use warp::{filters::BoxedFilter, path::Tail, Filter, Rejection, Reply};
+
+#[derive(Copy, Clone, Debug)]
+enum Error {
+    IndexHtmlMissing,
+    PathConversionFail,
+}
+
+#[derive(Serialize)]
+struct ErrorMessage {
+    code: u16,
+    message: String,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.description())
+    }
+}
+
+impl From<Error> for Rejection {
+    fn from(err: Error) -> Self {
+        warp::reject::custom(err)
+    }
+}
+
+impl StdError for Error {
+    fn description(&self) -> &str {
+        match self {
+            Error::IndexHtmlMissing => "index.html file not found",
+            Error::PathConversionFail => "Conversion of the path to PathAndQuery failed",
+        }
+    }
+
+    fn cause(&self) -> Option<&dyn StdError> {
+        None
+    }
+}
+
+pub fn create() -> BoxedFilter<(impl Reply,)> {
+    warp::any()
+        .and(warp::path::tail())
+        .and_then(serve)
+        .recover(unpack_problem)
+        .boxed()
+}
+
+fn serve(path: Tail) -> Result<impl Reply, Rejection> {
+    let path = path.as_str();
+    let path_and_query = PathAndQuery::from_str(path).map_err(|e| {
+        error!("Could not convert path {} to PathAndQuery: {:?}", path, e);
+        Error::PathConversionFail
+    })?;
+    let path = path_and_query.path();
```

Comment:
:confused: How many times do you want to convert back and forth between `Path`s? :see_no_evil: 

Is there no easier way to do this?

The function `guess_mime_type_opt` [takes anything that is `AsRef<Path>`](https://docs.rs/mime_guess/2.0.0-alpha.6/mime_guess/fn.guess_mime_type_opt.html), which is implemented for `&str` if I remember correctly. You should be able to simply pass `path.as_str()` to it.

Also, it looks like using [`warp::path::full()`](https://docs.rs/warp/0.1.15/warp/filters/path/struct.FullPath.html) might be more appropriate than `warp::path::tail()` because you actually want to match against the full thing and not the remaining, unmatched part.
In your concrete case, it is semantically equivalent because you don't do any matching before.

Reply:
> 😕 How many times do you want to convert back and forth between Paths? 🙈
I am sure you can find a better way to phrase your point.

I'll tackle the rest of the comment.

Reply:
Please note that unfortunately warp hides the "PathAndQuery" from the `Tail` struct making this cumbersome.

I'll see if I can make it better with `warp::path::full()`